service/frontend/adminHandler.go

```diff
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
+	run, err := adh.GetSDKClient().ExecuteWorkflow(
+		ctx,
+		client.StartWorkflowOptions{
+			TaskQueue: addsearchattributes.TaskQueueName,
+			ID:        addsearchattributes.WorkflowName,
+		},
+		addsearchattributes.WorkflowName,
+		wfParams,
+	)
 	if err != nil {
-		return nil, adh.error(serviceerror.NewInternal(fmt.Sprintf(errFailedUpdateDynamicConfigMessage, err)), scope)
+		return nil, adh.error(serviceerror.NewInternal(fmt.Sprintf(errUnableToStartWorkflowMessage, addsearchattributes.WorkflowName, err)), scope)
 	}
 
-	// update elasticsearch mapping, new added field will not be able to remove or update
-	index := adh.ESConfig.GetVisibilityIndex()
-	for k, v := range searchAttr {
-		esType := searchattribute.MapESType(v)
-		if len(esType) == 0 {
-			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errUnknownValueTypeMessage, v)), scope)
-		}
-		_, err := adh.ESClient.PutMapping(ctx, index, searchattribute.Attr, map[string]string{k: esType})
-		if err != nil {
-			return nil, adh.error(serviceerror.NewInternal(fmt.Sprintf(errFailedToUpdateESMappingMessage, err)), scope)
+	// Wait for workflow to complete.
+	err = run.Get(ctx, nil)
+	if err != nil {
+		scope.IncCounter(metrics.AddSearchAttributesWorkflowFailuresCount)
+		return nil, adh.error(serviceerror.NewInternal(fmt.Sprintf(errWorkflowReturnedErrorMessage, addsearchattributes.WorkflowName, err)), scope)
+	}
+	scope.IncCounter(metrics.AddSearchAttributesWorkflowSuccessCount)
+
+	// Get current state and return it.
+	resp, err := adh.getSearchAttributes(ctx, indexName, run.GetRunID())
+	if err != nil {
+		// Don't return error in case of getSearchAttributes error. Just log it.
+		adh.GetLogger().Warn("Unable to get search attributes back while adding them.", tag.Error(err))
+		metricsGetSearchAttributesScope := adh.GetMetricsClient().Scope(metrics.AdminGetSearchAttributesScope)
+		metricsGetSearchAttributesScope.IncCounter(metrics.ServiceFailures)
+	}
+
+	return &adminservice.AddSearchAttributesResponse{
+		CustomAttributes:         resp.GetCustomAttributes(),
+		SystemAttributes:         resp.GetSystemAttributes(),
+		Mapping:                  resp.GetMapping(),
+		AddWorkflowExecutionInfo: resp.GetAddWorkflowExecutionInfo(),
+	}, nil
+}
+
+func (adh *AdminHandler) GetSearchAttributes(ctx context.Context, request *adminservice.GetSearchAttributesRequest) (_ *adminservice.GetSearchAttributesResponse, retError error) {
+	defer log.CapturePanic(adh.GetLogger(), &retError)
+
+	scope, sw := adh.startRequestProfile(metrics.AdminGetSearchAttributesScope)
+	defer sw.Stop()
+
+	indexName := request.GetIndexName()
+	if indexName == "" {
+		indexName = adh.ESConfig.GetVisibilityIndex()
+	}
+
+	resp, err := adh.getSearchAttributes(ctx, indexName, "")
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	return resp, nil
+}
+
+func (adh *AdminHandler) getSearchAttributes(ctx context.Context, indexName, runID string) (*adminservice.GetSearchAttributesResponse, error) {
```

Comment:
`ctx context.Context, indexName, runID string` -> `ctx context.Context, indexName string, runID string`


let us try not to omit the type of a input var
` indexName, runID string` is hard to parse, compared to ` indexName string, runID string`